tests/test_pass_to_arg_and_types_str.py

```diff
@@ -0,0 +1,67 @@
+from dataclasses import dataclass, field
+from typing import Literal
+
+import pytest
+
+from xdsl.dialects import builtin
+from xdsl.ir import MLContext
+from xdsl.passes import ModulePass, get_pass_argument_names_and_types
+
+
+@dataclass
+class CustomPass(ModulePass):
+    name = "custom-pass"
+
+    number: int | float
+
+    single_number: int
+
+    int_list: list[int]
+
+    non_init_thing: int = field(init=False)
+
+    str_thing: str
+
+    nullable_str: str | None
+
+    literal: Literal["yes", "no", "maybe"] = "no"
+
+    optional_bool: bool = False
+
+    def apply(self, ctx: MLContext, op: builtin.ModuleOp) -> None:
+        pass
+
+
+@dataclass
+class EmptyPass(ModulePass):
+    name = "empty"
+
+    def apply(self, ctx: MLContext, op: builtin.ModuleOp) -> None:
+        pass
+
+
+@dataclass
+class SimplePass(ModulePass):
+    name = "simple"
+
+    a: int | float
+    b: int | None
+    c: int = 5
+
+    def apply(self, ctx: MLContext, op: builtin.ModuleOp) -> None:
+        pass
+
+
+@pytest.mark.parametrize(
+    "str_arg, pass_arg",
+    (
+        (
+            """number=int | float single_number=int int_list=list[int] non_init_thing=int str_thing=str nullable_str=str | None literal=no optional_bool=false""",
```

Comment:
A couple of nits remain:
1. the arguments are space-separated, but then so are the types in a union, which makes it more difficult to read, I find. Could you please change it so that its `int|float`?
2. The literal is also a little confusing, because all the other things are types, but `no` is not a type, while `Literal["no"]` is, could we print that somehow? If it's too much work then this seems lower priority, but would be good to file an issue to track it in the future.

Reply:
Okay will address.

Reply:
@math-fehr I managed to change the repr function by deleting the spaces around the stroke key | in the join. What do you think?

Reply:
I personally think that this is a bit ugly to be fair :/
`int|float` is much less readable than `int | float`, especially since it is always printed as `int | float` in "regular" python.

Reply:
@superlopuh What is your take? Happy with either - not an issue for me. 

For point 2): I raised an issue https://github.com/xdslproject/xdsl/issues/1845. Also added it to the file as a note.